beacon-chain/sync/service.go

```diff
@@ -189,6 +254,187 @@ func (ss *Service) ReceiveActiveState(data *pb.ActiveStateResponse) error {
 	return nil
 }
 
+// GetCrystallizedStateFromPeer sends a request to a peer for the corresponding crystallized state
+// for a beacon block.
+func (ss *Service) GetCrystallizedStateFromPeer(data *pb.BeaconBlockResponse, peer p2p.Peer) error {
+	block, err := types.NewBlock(data)
+	if err != nil {
+		return fmt.Errorf("could not instantiate new block from proto: %v", err)
+	}
+	h := block.CrystallizedStateHash()
+
+	log.Debugf("Successfully processed incoming block with crystallized state hash: %x", h)
+	ss.p2p.Send(&pb.CrystallizedStateRequest{Hash: h[:]}, peer)
+	ss.setStateMapping(h, block)
+	return nil
+}
+
+// SetFinalizedEpochFromCrystallizedState receives the crystallized state response from the peer
+// and sets the finalized epoch for the beacon block.
+func (ss *Service) SetFinalizedEpochFromCrystallizedState(data *pb.CrystallizedStateResponse) error {
+	state := types.NewCrystallizedState(data)
+
+	h, err := state.Hash()
+	if err != nil {
+		return fmt.Errorf("could not hash crystallized state: %v", err)
+	}
+	if ss.chainService.ContainsCrystallizedState(h) {
+		log.WithFields(logrus.Fields{"crystallizedStateHash": h}).Debug("Crystallized state hash exists locally")
+		return nil
+	}
+	log.Debugf("Successfully received incoming crystallized state with hash: %x", h)
+
+	finalizedBlock, ok := ss.stateMapping[h]
+	if !ok {
+		if err := ss.chainService.ProcessCrystallizedState(state); err != nil {
+			return fmt.Errorf("could not process crystallized state: %v", err)
+		}
+		return fmt.Errorf("unable to retrieve finalized block from mapping for hash: %x", h)
+	}
+
+	ss.setFinalizedEpochforMapping(h, finalizedBlock.BeaconBlock, state.LastFinalizedEpoch(), state)
+	log.Debugf("Saved finalized epoch for block with crystallised state hash: %x", h)
+	return nil
+}
+
+// validateAndSaveNextBlock will validate whether blocks received from the blockfetcher
+// routine can be added to the chain.
+func (ss *Service) validateAndSaveNextBlock(data *pb.BeaconBlockResponse) error {
+	block, err := types.NewBlock(data)
+	if err != nil {
+		return fmt.Errorf("could not instantiate new block from proto: %v", err)
+	}
+
+	if ss.currentSlotNumber == uint64(0) {
+		return fmt.Errorf("invalid slot number for syncing")
+	}
+
+	if (ss.currentSlotNumber + 1) == block.SlotNumber() {
+
+		if err := ss.writeBlockToDB(block); err != nil {
+			return err
+		}
+		ss.currentSlotNumber = block.SlotNumber()
+	}
+
+	return nil
+}
+
+// findLatestFinalizedBlock will retrieve the initial finalized block to be
+// saved, after the inital block has been saved syncing will be carried out by
+// a separate routine(not implemented yet).
+func (ss *Service) findAndSaveLatestFinalizedBlock() error {
+
+	sync, err := ss.isFirstSync()
+
+	if err != nil {
+		return err
+	}
+
+	if !sync {
+		return nil
+	}
+
+	if len(ss.stateMapping) > 20 {
```

Comment:
can you explain where this 20 is coming from?

Reply:
20 is an arbitrary number , it could also be 10 or 5 , etc. currently we receive any randomly broadcasted block from the network, in order to start our sync it would be best if we could start it with a block with the largest finalized epoch. 
- We receive n blocks from the network 
- We then try to receive the corresponding crystallized state for those blocks 
- Once that is done we can then choose the block with the largest finalized epoch and start our sync from there 
In this case n = 20 , but it could be something else. Ideally when we connect with a new peer we can have a handshake where we exchange information regarding our local chain, `LastFinalizedEpoch`, `slotNumber` , etc. Since we do not have this now , the above is one way we can start the sync. 

What are your thoughts do you think there is a better way to do this ?